engine/src/witness/sol/program_swaps_witnessing.rs

```diff
@@ -0,0 +1,508 @@
+};
+use futures::{stream, StreamExt, TryStreamExt};
+use itertools::Itertools;
+use state_chain_runtime::chainflip::solana_elections::SolanaVaultSwapDetails;
+use tracing::warn;
+
+const MAXIMUM_CONCURRENT_RPCS: usize = 16;
+// Querying less than 100 (rpc call max) as those event accounts can be quite big.
+// Max length ~ 1300 bytes per account. We set it to 10 as an arbitrary number to
+// avoid large queries.
+const MAX_MULTIPLE_EVENT_ACCOUNTS_QUERY: usize = 10;
+
+// 1. Query the on-chain list of opened accounts from SwapEndpointDataAccount.
+// 2. Check the returned accounts against the SC opened_accounts. The SC is the source of truth for
+//    the opened channels we can rely on that to not query the same accounts multiple times.
+// 3. If they are already seen in the SC we do nothing with them and skip the query.
+// 4. If an account is in the SC but not see in the engine we report it as closed.
+// 5. If they are not seen in the SC we query the account data. Then we parse the account data and
+//    ensure it's a valid a program swap. The new program swap needs to be reported to the SC.
+
+pub async fn get_program_swaps(
+	sol_rpc: &SolRetryRpcClient,
+	swap_endpoint_data_account_address: SolAddress,
+	sc_open_accounts: Vec<SolAddress>,
+	sc_closure_initiated_accounts: BTreeSet<VaultSwapAccountAndSender>,
+	usdc_token_mint_pubkey: SolAddress,
+) -> Result<
+	(Vec<(VaultSwapAccountAndSender, SolanaVaultSwapDetails)>, Vec<VaultSwapAccountAndSender>),
+	anyhow::Error,
+> {
+	let (new_program_swap_accounts, closed_accounts, slot) = get_changed_program_swap_accounts(
+		sol_rpc,
+		sc_open_accounts,
+		sc_closure_initiated_accounts,
+		swap_endpoint_data_account_address,
+	)
+	.await?;
+
+	if new_program_swap_accounts.is_empty() {
+		return Ok((vec![], closed_accounts));
+	}
+
+	let new_swaps = stream::iter(new_program_swap_accounts)
+		.chunks(MAX_MULTIPLE_EVENT_ACCOUNTS_QUERY)
+		.map(|new_program_swap_accounts_chunk| {
+			get_program_swap_event_accounts_data(sol_rpc, new_program_swap_accounts_chunk, slot)
+		})
+		.buffered(MAXIMUM_CONCURRENT_RPCS)
+		.map_ok(|program_swap_account_data_chunk| {
+			stream::iter(program_swap_account_data_chunk.into_iter().filter_map(
+				|(account, program_swap_account_data)| match program_swap_account_data {
+					Some(data)
+						if (data.src_token.is_none() ||
+							data.src_token.is_some_and(|addr| addr == usdc_token_mint_pubkey.into())) => {
+
+								let (deposit_metadata, vault_swap_parameters) = match data.ccm_parameters {
+									None => {
+										let CfParameters { ccm_additional_data: (), vault_swap_parameters } =
+											CfParameters::decode(&mut &data.cf_parameters[..]).map_err(|e| warn!("error while decoding CfParameters for solana vault swap: {}. Omitting swap", e)).ok()?;
+										(None, vault_swap_parameters)
+									},
+									Some(ccm_parameters) => {
+										let CfParameters { ccm_additional_data, vault_swap_parameters } =
```

Comment:
Just for clarity's sake
```suggestion
										let CcmCfParameters { ccm_additional_data, vault_swap_parameters } =
```